[mlir][memref][transform] Add new alloca_to_global op.

ingomueller-net opened the pull request:
This PR adds a new transform op that replaces `memref.alloca`s with `memref.get_global`s to newly inserted `memref.global`s. This is useful, for example, for allocations that should reside in the shared memory of a GPU, which have to be declared as globals.

ingomueller-net committed:
[mlir][memref][transform] Add new alloca_to_global op.

This PR adds a new transform op that replaces `memref.alloca`s with
`memref.get_global`s to newly inserted `memref.global`s. This is useful,
for example, for allocations that should reside in the shared memory of
a GPU, which have to be declared as globals.

ingomueller-net commented:
I am not sure whether the `module` argument is really necessary. I do need access to the surrounding module to insert the `memref.global`s, but I can do that through code (with `getParentOfType`). Also, I leave the module itself (and all other ops exact for the `alloca` inputs) intact and only add new ones.

I had an original version with that argument and that ran, except for some crashes that I retrospectively relate to #66357. It may, thus, work without the argument but is it legal to do so?

ingomueller-net commented:
> I am not sure whether the `module` argument is really necessary. I do need access to the surrounding module to insert the `memref.global`s, but I can do that through code (with `getParentOfType`). Also, I leave the module itself (and all other ops exact for the `alloca` inputs) intact and only add new ones.
> 
> I had an original version with that argument and that ran, except for some crashes that I retrospectively relate to #66357. It may, thus, work without the argument but is it legal to do so?

@ftynse: Can you help us with this? @nicolasvasilache and @matthias-springer seemed to say that this might be safe but weren't very confident in their assessment...

ftynse commented on mlir/include/mlir/Dialect/MemRef/TransformOps/MemRefTransformOps.td:
```diff
@@ -144,6 +144,71 @@ def ApplyResolveRankedShapedTypeResultDimsPatternsOp : Op<Transform_Dialect,
+    `memref.global` ops that were inserted by the transformation.
+  }];
+
+  let arguments = (ins Transform_ConcreteOpType<"builtin.module">:$module,
```
I don't think we want to limit this argument to be specifically `builtin.module`. In the use case the documentation describes, this can clearly also be a `gpu.module`. But in general it is any symbol table.

ingomueller-net replied:
OK, good point.

Coming back to the question of whether we need this argument at all: `memref.get_global` expects the symbol to be in the nearest `SymbolTable` (see [verifier](https://github.com/llvm/llvm-project/blob/13e9a568/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp#L1635-L1639)). This means that, given the `memref.alloca` payload, there is only one symbol table to insert the symbol into. Do we really need to pass in that op explicitly, then?

ingomueller-net replied:
Allowing any op with the `SymbolTable` trait in the newest revision.

ftynse replied:
>  Do we really need to pass in that op explicitly, then?

I can imagine a hypothetical future where we can refer to globals through nested named scopes. But it's not the case today, so both ways are fine.

ftynse commented on mlir/lib/Dialect/MemRef/TransformOps/MemRefTransformOps.cpp:
```diff
@@ -126,6 +126,96 @@ void transform::ApplyResolveRankedShapedTypeResultDimsPatternsOp::
+// AllocaToGlobalOp
+//===----------------------------------------------------------------------===//
+
+namespace {
+static llvm::SmallString<64> getUniqueSymbol(llvm::StringRef prefix,
+                                             ModuleOp module) {
+  llvm::SmallString<64> candidateNameStorage;
+  StringRef candidateName(prefix);
+  int uniqueNumber = 0;
+  while (true) {
+    if (!module.lookupSymbol(candidateName)) {
+      break;
+    }
+    candidateNameStorage.clear();
+    candidateName = (prefix + Twine("_") + Twine(uniqueNumber))
+                        .toStringRef(candidateNameStorage);
+    uniqueNumber++;
+  }
+  return candidateName;
+}
+} // namespace
```
SymbolTable should have support for symbol uniquing. Let's use that.

ingomueller-net replied:
Done in the newest revision.

ftynse commented on mlir/lib/Dialect/MemRef/TransformOps/MemRefTransformOps.cpp:
```diff
@@ -126,6 +126,96 @@ void transform::ApplyResolveRankedShapedTypeResultDimsPatternsOp::
+      rewriter.setInsertionPointToStart(&module.getBodyRegion().front());
+      Type resultType = alloca.getResult().getType();
+      llvm::SmallString<64> symName = getUniqueSymbol("alloca", module);
+      // XXX: Add a better builder for this.
```
Nit: XXX->TODO

ingomueller-net replied:
Done in the newest revision.

ftynse commented on mlir/lib/Dialect/MemRef/TransformOps/MemRefTransformOps.cpp:
```diff
@@ -126,6 +126,96 @@ void transform::ApplyResolveRankedShapedTypeResultDimsPatternsOp::
+    auto getGlobalOp = rewriter.replaceOpWithNewOp<memref::GetGlobalOp>(
+        alloca, globalOp.getType(), globalOp.getName());
+
+    globalOps.push_back(globalOp);
+    getGlobalOps.push_back(getGlobalOp);
```
Nit: could we `reserve` before pushing to vector in a loop?

ingomueller-net replied:
Done in the latest revision. Note, though, that looking up the number of alloca ops isn't a constant-time operation and calling `llvm::range_size(allocaOps)` could theoretically be more expensive than the `push_back`s.

ftynse replied:
The previous version was doing it anyway IIRC. If we only need it for reserve, then let's drop the reserve.

ftynse commented on mlir/test/python/dialects/transform_memref_ext.py:
```diff
@@ -16,6 +16,54 @@ def run(f):
+    sequence = transform.SequenceOp(
+        transform.FailurePropagationMode.Propagate,
+        [],
+        transform.OperationType.get("memref.alloc"),
+    )
+    with InsertionPoint(sequence.body):
+        module = transform.CastOp(
+            transform.OperationType.get("builtin.module"), sequence.bodyTarget
+        )
+        alloca = transform.CastOp(
+            transform.OperationType.get("memref.alloca"), sequence.bodyTarget
+        )
+        memref.MemRefAllocaToGlobalOp(module, alloca)
+        transform.YieldOp()
```
I understand it's a test, but types looks quite wrong here. We are casting `op<"memref.alloc">` to two different other concrete types. This cannot be valid and may break should we introduce some static "type compatibility" checks in the cast. Sequence can have multiple block arguments, let's just use that and give them the expected type immediately.

ingomueller-net replied:
Yep, much better, thanks! Fixed in latest revsion.

ingomueller-net committed:
Address comments from @ftynse's review.

In particular:
* Accept any op type with `SymbolTable` trait as containing op rather
  than only `builtin.module` and rename op argument accordingly.
* Use `SymbolTable::insert` to unique the name of the globals rather
  than some hand-rolled function.
* Use more sane semantics in Python mix-in test.

ingomueller-net committed:
Remove symbolTable op arg and simplify tests.

ingomueller-net committed:
Remove symbolTable op arg in Python mix-in.
